text/0000-pack-buildpack-new-alternatives.md

```diff
@@ -0,0 +1,210 @@
+# Meta
+[meta]: #meta
+- Name: Pack Buildpack New to Support Alternatives
+- Start Date: 2022-03-10
+- Author(s): aidandelaney
+- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+The `pack buildpack new` subcommand creates a new buildpack based on a shell-script template.  This proposal extends `pack buildpack new` with a `--libcnb` option and a `--template URL` option to allow alternatives to the shell-script template.  The `--libcnb` option creates a new buildpack written in golang using `libcnb`.  The `--template URL` option allows buildpack creation from an arbitrary 3rd party URL.
+
+# Definitions
+[definitions]: #definitions
+
+* **project scaffolding**: minimal code and file system structure used to start a project.
+
+# Motivation
+[motivation]: #motivation
+
+The creation of buildpacks in languages other than `bash` is undocumented in the [Buildpack Author Guide](https://buildpacks.io/docs/buildpack-author-guide).  In particular, creating a buildpack using `libcnb` is undocumented because there is no mechanism to generate a scaffolded project.  This proposal adds a minimal `libcnb` project template to `pack` and allows 3d parties to provide templates to create buildpacks for their chosen technology stack.
+
+Extending `pack buildpack new` with a `--libcnb` flag supports buildpack creation using the existing documented workflow.  This results easier on-boarding of buildpack authors into the `libcnb` ecosystem.  Additionally extending `pack buildpack new` with `--target URL` allows other projects to re-use `pack` as their scaffolding tool.
+
+`pack buildpack new --libcnb` supports end-users who wish to scaffold a new `libcnb`-based buildpack.  The operation of `pack buildpack new --libcnb` does not require internet access to succeed.  `pack buildpack new --template URL` allows projects (such as Paketo, Heroku and others) to define skeleton projects for new buildpacks.
+
+Extending `pack buildpack new` allows buildpack authors to more-easily create new buildpacks in their chosen language and framework.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Project scaffolding is [very](https://cookiecutter.readthedocs.io/) [popular](https://yeoman.io/) [in](https://github.com/kowainik/summoner) [other](https://crates.io/crates/cargo-scaffold) [ecosystems](https://github.com/golang-standards/project-layout).  Scaffolding systems are employed to ease onboarding of new developers.  Within `pack` this feature is targeted at onboarding buildpack authors.
+
+- Explaining the feature largely in terms of examples.
```

Comment:
What should the behavior be when no opts are provided (i.e. just `pack buildpack new`)? Does it still default to bash? Should we prompt with a choice of language/template?

Reply:
I'm very open to other choices, but I was thinking we should preserve the current behaviour with `pack buildpack new`.  However, prompting the end-user for a choice of implementation language is something we should discuss.

Reply:
Yea, I'm open to a lot of options too. I like bash as the default because it's a kind of lowest common denominator (and we can even add https://github.com/jkutner/libcnb.bash if want). But you're right that it implies something about our project's recommendations.

Reply:
If no parameters are provided, end users are prompted to choose from the default template collection.  A prototype collection looks like https://github.com/AidanDelaney/cnb-buildpack-templates.